From 7dd611828bb181c29d20544c7dcae8f52215ae70 Mon Sep 17 00:00:00 2001
From: Paul Gofman <gofmanp@gmail.com>
Date: Thu, 9 Jan 2020 15:05:09 +0300
Subject: [PATCH] ntdll: Stop search on mmap() error in try_map_free_area().

The anon mmap errors do not depend on start address hint. Ignoring them
makes the search take incredible time until it fails.
---
 dlls/ntdll/virtual.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/virtual.c b/dlls/ntdll/virtual.c
index c6f3c1685e0..37f83efa8fc 100644
--- a/dlls/ntdll/virtual.c
+++ b/dlls/ntdll/virtual.c
@@ -767,8 +767,14 @@ static void* try_map_free_area( void *base, void *end, ptrdiff_t step,
             return start;
         TRACE( "Found free area is already mapped, start %p.\n", start );
 
-        if (ptr != (void *)-1)
-            munmap( ptr, size );
+        if (ptr == (void *)-1)
+        {
+            ERR("wine_anon_mmap() error %s, start %p, size %p, unix_prot %#x.\n",
+                    strerror(errno), start, (void *)size, unix_prot);
+            return NULL;
+        }
+
+        munmap( ptr, size );
 
         if ((step > 0 && (char *)end - (char *)start < step) ||
             (step < 0 && (char *)start - (char *)base < -step) ||
-- 
2.26.2

